src/devinterp/utils.py

```diff
@@ -1,31 +1,91 @@
+from torch.utils.data import Dataset, DataLoader
+import torch
 import numpy as np
 import matplotlib.pyplot as plt
 
 
-def plot_trace(trace, y_axis, x_axis='step', title=None, plot_mean=True, plot_std=True, fig_size=(12, 9), true_lc = None):
+def plot_trace(
+    trace,
+    y_axis,
+    x_axis="step",
+    title=None,
+    plot_mean=True,
+    plot_std=True,
+    fig_size=(12, 9),
+    true_lc=None,
+):
     num_chains, num_draws = trace.shape
     sgld_step = list(range(num_draws))
     if true_lc:
-        plt.axhline(y=true_lc, color='r', linestyle='dashed')
+        plt.axhline(y=true_lc, color="r", linestyle="dashed")
     # trace
     for i in range(num_chains):
         draws = trace[i]
-        plt.plot(sgld_step, draws, linewidth=1, label=f'chain {i}')
+        plt.plot(sgld_step, draws, linewidth=1, label=f"chain {i}")
 
     # mean
     mean = np.mean(trace, axis=0)
-    plt.plot(sgld_step, mean, color='black', linestyle='--', linewidth=2, label='mean', zorder=3)
-    
+    plt.plot(
+        sgld_step,
+        mean,
+        color="black",
+        linestyle="--",
+        linewidth=2,
+        label="mean",
+        zorder=3,
+    )
+
     # std
     std = np.std(trace, axis=0)
-    plt.fill_between(sgld_step, mean - std, mean + std, color='gray', alpha=0.3, zorder=2)
+    plt.fill_between(
+        sgld_step, mean - std, mean + std, color="gray", alpha=0.3, zorder=2
+    )
 
     if title is None:
-        title = f'{y_axis} values over sampling draws'
-    plt.legend(loc='upper left', bbox_to_anchor=(1, 1))    
+        title = f"{y_axis} values over sampling draws"
+    plt.legend(loc="upper left", bbox_to_anchor=(1, 1))
     plt.title(title)
     plt.xlabel(x_axis)
     plt.ylabel(y_axis)
     plt.figure(figsize=fig_size)
     plt.tight_layout()
     plt.show()
+
+
+def optimal_temperature(data):
```

Comment:
i think in practice, jesse and i both used the batch size for `num_samples`, `len(data.dataset)` was too large (and also didn't make sense to me). `batch_size` makes sense to me since varying the batch size does change the effective temperature of SGLD steps, but you should be able to e.g. run sampling using a dataset of 1m rows where you don't use all the rows and get the same llc estimate as a counterfactual sampling run using a dataset of 100m rows where you also don't use all the rows (but do the same batch size, num draws, num chains, etc)